### PR TITLE
ETS-514 | validate PHAC key fields

### PR DESCRIPTION
**Pull request by sunghwan-park-fw**
Refer to [ETS-514](https://freshworks.atlassian.net/browse/ETS-514)

### CODE REVIEW

File: server/validation.js [hunk 1]
```diff
@@ -44,6 +51,11 @@ const DeterminationSchema = yup.object().noUnknown().shape({
 const PhacSchema = yup.array().required('Submission must contain at least one item').of(
   yup.object().shape({
     covid_id: yup.string().typeError('covid_id must be a string').required('covid_id is required'),
+    date_of_birth: yup.string().required('Date of birth is required').test('is-date', 'Date of birth is invalid', validatePastDateString),
```
Comment (hitankar-freshworks) [inline server/validation.js:54]: I think we might want to validate the structure. Asked @fwkendall to verify

```
{
        "covid_id": "CVR-test1",
        "first_name": "Test",
        "last_name": "UserMcGee",
        "date_of_birth": "1967-01-01",
        "type": "Asymptomatic",
        "arrival_date": "2020-01-01",
        "end_of_isolation": "2020-01-15",
        "destination_type": "Home",
        "address_1": "",
        "city": "",
        "postal_code": "",
        "home_phone": "555-555-5555",
        "mobile_phone": "",
        "other_phone": "",
        "email_address": "",
        "port_of_entry": "Land border",
        "province_territory": "British Columbia"
    }
```

File: server/validation.js [hunk 2]
```diff
@@ -44,6 +51,11 @@ const DeterminationSchema = yup.object().noUnknown().shape({
...
+    arrival_date: yup.string().required('Arrival date is required').test('is-date', 'Arrival date is invalid', validateDateString),
```
Comment (hitankar-freshworks) [inline server/validation.js:55]: Wonder we should do simple check if arrival date is after dob.

File: server/validation.js (hunk 1 above)
Comment (hitankar-freshworks) [inline server/validation.js:54]: Can `dob` have min value: `1900-01-01` and max of new Date()

File: server/validation.js (hunk 2 above)
Comment (hitankar-freshworks) [reply server/validation.js:55]: `end_of_isolation` needs to be validated and is probably a date after arrival

File: server/validation.js (hunk 1 above)
Comment (hitankar-freshworks) [reply server/validation.js:54]: We probably want postal code validation. This regex worked well in the past:
`[ABCEGHJKLMNPRSTVXY]\d[ABCEGHJ-NPRSTV-Z][\s-]?\d[ABCEGHJ-NPRSTV-Z]\d`

File: server/validation.js [hunk 3]
```diff
@@ -36,6 +36,13 @@ const validateUniqueArray = (a) => (
   Array.isArray(a) && new Set(a).size === a.length
 );
 
+const PHONE_NUMBER_PATTERN = /^((\+[1-9]{1,4}[ \\-]*)|(\([0-9]{2,3}\)[ .\\-]*)|([0-9]{2,4})[ .\\-]*)*?[0-9]{3,4}?[ .\\-]*[0-9]{3,4}?$/;
```
Comment (hitankar-freshworks) [inline server/validation.js:39]: What format of numbers is being validated?

**Comment by hitankar-freshworks** [comment]
I think I am missing the part where were validate method is called. Is the validator already been called in the API?

File: server/tests/phac-ets-functions.test.js [hunk 4]
```diff
@@ -9,6 +9,8 @@ const { fromCsvString } = require('./util/csv');
 const markDuplicates = require('../lambda/phacToSbc/mark-duplicates');
 const { sendPhacToSBC, sendEtsToSBC } = require('../lambda/layer/common/nodejs/custom_modules/send-to-sbc');
 
+jest.setTimeout(600000);
```
[the remark this answers is not in the document]
Comment (sunghwan-park-fw) [reply server/tests/phac-ets-functions.test.js:12]: it should be deleted.

File: server/validation.js (hunk 1 above)
Comment (hitankar-freshworks) [reply server/validation.js:54]:
```
{
        "covid_id": "CVR-test1", 
        "first_name": "Test", 
        "last_name": "UserMcGee",
        "date_of_birth": "1967-01-01", // validate date
        "type": "Asymptomatic",
        "arrival_date": "2020-01-01", // required
        "end_of_isolation": "2020-01-15", // required
        "destination_type": "Home",
        "address_1": "",
        "city": "",
        "postal_code": "",
        "home_phone": "555-555-5555", // check format
        "mobile_phone": "", // check format
        "other_phone": "", // check format
        "email_address": "", // check format
        "port_of_entry": "Land border",
        "province_territory": "British Columbia" // required
    }
```

Comment (hitankar-freshworks) [reply server/validation.js:54]: We can validate the required fields in the above list. Others marked as `check format` need to be validated based on data from the database.

File: server/package.json [hunk 5]
```diff
@@ -28,7 +28,7 @@
     "passport-local": "1.0.0",
     "winston": "3.3.3",
     "winston-cloudwatch": "0.9.0",
-    "yup": "0.29.3"
```
Comment (sunghwan-park-fw) [inline server/package.json:31]: this version doesn't pass context to test function.

File: server/validation.js [hunk 6]
```diff
@@ -52,10 +54,40 @@ const PhacSchema = yup.array().required('Submission must contain at least one it
...
+      if (Date.parse(value) < Date.parse(parent.arrival_date)) {
+        const message = `End of isolation should be later than arrival date: ${value} < ${parent.arrival_date}`;
+        return createError({ path, message });
+      }
+      return true;
+    }),
+    home_phone: yup.string().test('is-phone-number', 'Home phone number is invalid', validatePhoneNumber),
+    mobile_phone: yup.string().test('is-phone-number', 'Mobile phone number is invalid', validatePhoneNumber),
+    other_phone: yup.string().when(['home_phone', 'mobile_phone'], {
+      is: (homePhone, mobilePhone) => !homePhone && !mobilePhone,
```
Comment (hitankar-freshworks) [inline server/validation.js:78]: Isnt this validation needed for `home_phone` and `mobile_phone` as well?

File: server/validation.js [hunk 7]
```diff
@@ -52,10 +54,40 @@ const PhacSchema = yup.array().required('Submission must contain at least one it
...
+    end_of_isolation: yup.string().required('End of isolation is required').test('is-date', 'End of isolation is invalid', (value, context) => {
+      if (!validateDateString(value)) return false;
+      const { path, parent, createError } = context;
+      if (Date.parse(value) < Date.parse(parent.arrival_date)) {
+        const message = `End of isolation should be later than arrival date: ${value} < ${parent.arrival_date}`;
+        return createError({ path, message });
+      }
+      return true;
+    }),
+    home_phone: yup.string().test('is-phone-number', 'Home phone number is invalid', validatePhoneNumber),
```
Comment (hitankar-freshworks) [inline server/validation.js:75]: what if the phone number is a number, would it fail this validation?

**Review by hitankar-freshworks — APPROVED** [review]
Two important comments. looks good otherwise.

**Pull request merged**